### PR TITLE
Update API

### PR DESCRIPTION
**Pull request by weitingchou**
1. Update database schema for allowing definition of subdocument
2. Renew API
3. Add brain implementation for communication with new API

### CODE REVIEW

File: services/api/analyzers.js [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+
+    function closeResponse() {
+        ignore = true
+        clearTimeout(timer)
+    }
+
+    nats.request(CH_API_BRAIN, request, {'max': NUM_OF_BRAINS}, (reply) => {
+        if (!ignore) {
+            count += 1
+            let isLastReply = count === NUM_OF_BRAINS ? true : false
```
Comment (SuJiaKuan) [inline services/api/analyzers.js]: Just use:
```javascript
const isLastReply = count === NUM_OF_BRAINS;
```

File: services/api/analyzers.js [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+                }
+            }
+            return next(createError(500, null, err))
+        }
+        res.status(201).send({id: saved.id})
+    })
+}
+
+function deleteAllAnalyzers(req, res, next) {
+
```
Comment (SuJiaKuan) [inline services/api/analyzers.js]: Please insert `TODO` comment.

File: services/api/analyzers.js [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+                    closeResponse()
+                    next(createError(404, `Runtime instance of analyzer "${id}" was not found`))
+                }
+            }
+        }
+    })
+}
+
+function updateAnalyzerConfig(req, res, next) {
+
```
Comment (SuJiaKuan) [inline services/api/analyzers.js]: Please insert `TODO` comment.

File: services/api/analyzers.js [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+        }
+        if (result === null) {
+            return next(createError(404))
+        }
+        res.send(result)
+    })
+}
+
+function updateAnalyzerSourceConfig(req, res, next) {
+
```
Comment (SuJiaKuan) [inline services/api/analyzers.js]: Please insert `TODO` comment.

File: services/api/analyzers.js [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+        }
+        if (result === null) {
+            return next(createError(404))
+        }
+        res.send(result)
+    })
+}
+
+function updateAnalyzerPipelineConfig(req, res, next) {
+
```
Comment (SuJiaKuan) [inline services/api/analyzers.js]: Please insert `TODO` comment.

File: services/api/analyzers.js [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+            if (reply['result']) {
+                closeResponse()
+                return res.status(201).send(reply['result'])
+            }
+        })
+    })
+}
+
+function updateAnalyzerRuntime(req, res, next) {
+
```
Comment (SuJiaKuan) [inline services/api/analyzers.js]: Please insert `TODO` comment.

File: services/api/analyzers.js [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+}
+
+function deleteAnalyzerRuntime(req, res, next) {
```
Comment (SuJiaKuan) [inline services/api/analyzers.js]: The function is almost the same with *deleteAnalyzer()*. It's recommended to write a reusable function for both of them (in next PR).

File: services/api/analyzers.js [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+            let isLastReply = count === NUM_OF_BRAINS ? true : false
+            if (isLastReply) {
+                // All replies are received, cancel the timeout
+                clearTimeout(timer)
+            }
+            try {
+                const replyJSON = JSON.parse(reply)
+                if (replyJSON['code'] &&
+                    replyJSON['code'] === msg['ch_api_brain_reply']['NOT_AVAILABLE']) {
+                    let errReply = {
```
Comment (SuJiaKuan) [inline services/api/analyzers.js]: Use `const` instead of `let`.

File: services/api/analyzers.js [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+                        error: {
+                            code: msg['ch_api_brain_reply']['NOT_AVAILABLE'],
+                            message: 'Runtime instance is not available to accept request right now'
+                        }
+                    }
+                    return cb(errReply, null, closeResponse)
+                }
+                cb(replyJSON, isLastReply, closeResponse)
+            } catch (e) {
+                let errReply = { error: { message: e } }
```
Comment (SuJiaKuan) [inline services/api/analyzers.js]: Use `const` instead of `let`.

File: services/api/analyzers.js [hunk 10]
```diff
@@ -0,0 +1,413 @@
...
+                const replyJSON = JSON.parse(reply)
+                if (replyJSON['code'] &&
+                    replyJSON['code'] === msg['ch_api_brain_reply']['NOT_AVAILABLE']) {
+                    let errReply = {
+                        error: {
+                            code: msg['ch_api_brain_reply']['NOT_AVAILABLE'],
+                            message: 'Runtime instance is not available to accept request right now'
+                        }
+                    }
+                    return cb(errReply, null, closeResponse)
```
Comment (SuJiaKuan) [inline services/api/analyzers.js]: Why is *null* fed into *isLastReply* for error case?

File: apps/tripwire/brain/brain.py [hunk 11]
```diff
@@ -260,85 +264,94 @@ def __init__(self, ch_public='public_brain',
...
+                req = {
+                    'command': MESSAGES['ch_brain_res']['CREATE_WORKER'],
+                    'ticketId': ticket_id,
+                    'params': {
+                        # TODO: For running multiple brain instances, the id
+                        #       should combine with a brain id to create a
+                        #       unique id across brains
+                        'workerName': 'jagereye/worker_tripwire'
                     }
-                    await self._mem_db_cli.set(ticket_id , str(ticket_obj))
```
Comment (uniray7) [inline apps/tripwire/brain/brain.py]: it needs to insert ticket in mem_db, right?

File: apps/tripwire/brain/contract.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        self._typename = typename
+        self._msg = MESSAGES['ch_api_brain']
+        self._msg_r = MESSAGES['ch_api_brain_reply']
+
+    def validate(self, request):
+        try:
+            if request['command'] == self._msg['START_ANALYZER']:
+                if self._typename != request['params']['type']:
+                    raise InvalidRequestType
+            if not request['params']['source'] or \
```
Comment (uniray7) [inline apps/tripwire/brain/contract.py]: is the indent right? it should be belong 'if request['command'] == self._msg['START_ANALYZER']: '  if-else statement

File: apps/tripwire/brain/contract.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+               not request['params']['pipelines']:
+                raise InvalidRequestFormat
+            if request['command'] == self._msg['REQ_ANALYZER_STATUS'] or \
```
Comment (uniray7) [inline apps/tripwire/brain/contract.py]: should use elif instead of if

File: apps/tripwire/brain/contract.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    pass
+
+
+class InvalidRequestFormat(Exception):
+    pass
+
+
+class API():
+    def __init__(self, typename):
+        self._typename = typename
```
Comment (uniray7) [inline apps/tripwire/brain/contract.py]: maybe we should discuss the naming 'typename'

**Pull request merged**